---
title: "Preprocessing and QC of simulated data set"
output: html_document
editor_options: 
  chunk_output_type: console
params:
  dataset: ""
---

```{r load-packages}
suppressPackageStartupMessages({
  #library(MultiAssayExperiment)
  library(SingleCellExperiment)
  library(scater)
  library(scran)
  library(plyr)
  library(dplyr)
  library(ggplot2)
  library(ComplexHeatmap)
  library(splatter)
})
```


## Read in simulated data and normalization

```{r simulate-counts}
dataset = params$dataset
simuname <- paste0("/home/xianingz/storage/Github/scRNAseqBenchmark_Clustering/Data/Simulations/",dataset)
sce <- readRDS(simuname)
sce <- normalise(sce, exprs_values = "counts", return_log = TRUE) ## generates logcounts(sce)
```

## Exclude features that are not expressed

```{r reduce-expression-matrix}
keep_features <- rowSums(counts(sce) > 0) > 0
table(keep_features)
sce <- sce[keep_features, ]
dim(sce)
```

## Calculate QC metrics

```{r QC}
sce <- calculateQCMetrics(sce)
```

## Quality control using PCA on column data

We create a PCA plot based the quality metrics for each cell, e.g., the total
number of reads and the total number of features.

```{r qc-pca}
sce <- scater::runPCA(sce, pca_data_input = "coldata")
scater::plotPCA(sce, colour_by = "Group")
```


## Filter cells

We remove cells with log-library sizes (or total features) that are more than 3
median absolute deviations (MADs) below the median log-library size (or total
features).

```{r histogram}
colData(sce)$libsize.drop <- isOutlier(sce$total_counts, nmads = 3, type = "lower", log = TRUE)
ggplot(as.data.frame(colData(sce)), aes(x = total_counts)) + 
  geom_histogram(bins = 20, fill = "grey80") + xlab("Total count") + 
  ylab("Number of cells") + 
  geom_vline(xintercept = min(sce$total_counts[!sce$libsize.drop]), 
             color = "red", linetype = "dashed") + 
  theme_bw()

colData(sce)$feature.drop <- isOutlier(sce$total_features_by_counts, nmads = 3, type = "lower", log = TRUE)
ggplot(as.data.frame(colData(sce)), aes(x = total_features_by_counts)) + 
  geom_histogram(bins = 20, fill = "grey80") + xlab("Number of detected features") + 
  ylab("Number of cells") + 
  geom_vline(xintercept = min(sce$total_features[!sce$feature.drop]), 
             color = "red", linetype = "dashed") + 
  theme_bw()
```

```{r filter-cells}
table(libsize = sce$libsize.drop, feature = sce$feature.drop)
sce <- sce[, !(sce$libsize.drop | sce$feature.drop)]
dim(sce)
```

## Quality control using highest expressed genes

```{r qc-filt}
plotQC(sce, type = "highest-expression", n = 50)
```

## Data normalization

```{r sizefactors}
sce <- computeSumFactors(sce, sizes = pmin(ncol(sce), seq(20, 120, 20)), min.mean = 0.1)
summary(sizeFactors(sce))
```

```{r normalization}
sce <- normalise(sce, exprs_values = "counts", return_log = TRUE)
sce <- normalise(sce, exprs_values = "counts", return_log = FALSE)
```

## Plot the proportion of explained variances

```{r explained-variance, warning = FALSE} 
expl_vars <- c("Group", "log10_total_counts", "log10_total_features_by_counts",
               "pct_counts_in_top_100_features")
plotExplanatoryVariables(sce, variables = expl_vars)
```

## Plot t-SNE representations

```{r tSNE}
set.seed(1234)
sce <- runTSNE(sce, exprs_values = "logcounts", perplexity = 10)
plotTSNE(sce, colour_by = "Group")
plotTSNE(sce, colour_by = "total_features_by_counts", size_by = "total_counts")
```

## Show cluster blocks using Heatmap

```{r heatmap}
sce.cor <- cor(normcounts(sce))
h=Heatmap(sce.cor[order(sce$Group),order(sce$Group)],cluster_rows = FALSE,cluster_columns = FALSE,show_row_names = FALSE,show_column_names = FALSE, name = "Correlation")
draw(h)
```


## Save the normalized and cell filtered dataset

```{r save-data}
dim(sce)
table(sce$Group)
saveRDS(sce, file = paste0("/home/xianingz/storage/Github/scRNAseqBenchmark_Clustering/Data/data_pre/",dataset))
```

## Session info

```{r}
date()
sessionInfo()
```
